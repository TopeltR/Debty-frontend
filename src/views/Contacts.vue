<template>
    <div>
        <navbar/>
        <background>
            <div class='search-wrapper pt-4'>
                <b-row class='ml-0'>
                    <h2>Add contact</h2>
                    <font-awesome-icon id="info" icon='info-circle' class='ml-1'></font-awesome-icon>
                    <b-tooltip target="info"
                               title="Create an Event by giving it an informative title and contacts, then add your contacts to it."
                               placement="bottom">
                    </b-tooltip>
                </b-row>
                <b-row class='pt-4'>
                    <b-col sm='12' md='6'>
                        <autocomplete id='search ' v-model='contact' :placeholder='"Name"' :field='field'
                                      :items='availableContacts'
                                      :key-extractor='getUserFullName'>
                        </autocomplete>
                    </b-col>
                    <b-col class='pt-3 pt-md-0'>
                        <button type='button' v-on:click='addContact'
                                class='btn btn-primary wide'> Add contact
                        </button>
                    </b-col>
                </b-row>
            </div>
            <b-row class='mt-5'>
                <b-col md='6'>
                    <b-list-group>
                        <h2>Requests</h2>
                        <div v-if='requests.length > 0'>
                            <b-list-group-item v-for='request in requests'> {{request.firstName}} {{request.lastName}}
                                <span v-if='request.type === "Incoming"'>
                                    <font-awesome-icon icon='check' class='color-green ml-1 icons float-right'
                                                       v-on:click='acceptContact(request.id)'/>
                                    <font-awesome-icon icon='times' class='color-red icons float-right' v-on:click='removeRequest(request.id)'/>
                                </span>
                                <div v-else></div>
                            </b-list-group-item>
                        </div>
                        <div v-else class='mt-3'>
                            <h4>No incoming requests right now.</h4>
                        </div>
                    </b-list-group>

                </b-col>
                <b-col md='6'>
                    <b-list-group>
                        <h2>My contacts</h2>
                        <div v-if="userContacts.length > 0">
                            <b-list-group-item v-for='contact in userContacts'> {{contact.firstName}} {{contact.lastName}}
                                <font-awesome-icon icon='times' class='color-red icons float-right' v-on:click='removeRequest(contact)'/>
                            </b-list-group-item>
                        </div>
                        <div v-else class='mt-3'>
                            <h3>No contacts yet, add contacts from search</h3>
                        </div>
                    </b-list-group>
                </b-col>
            </b-row>
        </background>
    </div>
</template>

<script>
    import Navbar from '@/components/Navbar.vue';
    import Background from '@/components/Background.vue';
    import Autocomplete from '@/components/Autocomplete.vue';
    import router from '@/router';
    import userStore from '../stores/UserStore';

    export default {
        name: 'events',
        components: {Navbar, Background, Autocomplete},
        data() {
            return {
                contact: {},
                field: {value: ''},
                availableContacts: [],
                requests: [],
                userContacts: [],
                user: userStore.getUser(),
            };
        },
        mounted() {
            this.getWaitingContacts();
            this.getContacts();
            this.getPersonContacts();
        },
        methods: {
            getUserFullName(user) {
                return user.lastName === null ? user.firstName : user.firstName + ' ' + user.lastName;
            },
            async getContacts() {
            this.user = await userStore.getUser();
            this.$http.get('/contact/all/' + this.user.id)
                .then((data) => {
                    this.availableContacts = data.data;
                }).catch(() => {
                    router.push('/');
                },
            );
            },
            async getWaitingContacts() {
                this.user = await userStore.getUser();
                this.$http.get('/contact/waiting/' + this.user.id)
                    .then((data) => {
                        this.requests = data.data;
                        this.requests.forEach((request) => {
                            Object.assign(request, {type: 'Incoming'});
                        });
                    }).catch(() => {
                        router.push('/');
                    },
                );
            },
            async addContact() {
                this.user = await userStore.getUser();
                this.$http.post('/contact/add/' + this.user.id + '/' + this.contact.id)
                    .then(this.getContacts).then(() => {
                    this.field.value = '';
                    this.requests.push(this.contact);
                })
                    .catch(() => {
                            router.push('/');
                        },
                    );
            },
            async getPersonContacts() {
                this.user = await userStore.getUser();
                this.$http.get('/contact/id/' + this.user.id)
                    .then((data) => {
                        this.userContacts = data.data;
                    }).catch(() => {
                        router.push('/');
                    },
                );
            },
            async acceptContact(id) {
                this.user = await userStore.getUser();
                this.$http.post('/contact/accept/' + this.user.id + "/" + id)
                    .then(() => {
                        this.getPersonContacts();
                        this.getWaitingContacts();
                    }).catch(() => {
                        router.push('/');
                    },
                );
            },
            async removeRequest(contact) {
                this.user = await userStore.getUser();
                this.$http.delete('/contact/remove/' + this.user.id + '/' + contact.id)
                    .then(() => {
                        this.getWaitingContacts();
                        this.getPersonContacts();
                    }).catch(() => {
                        router.push('/');
                    },
                );
            },
        },
    };
</script>

<style scoped>
    .search-wrapper {
        display: table;
        margin: 0 auto;
    }

    #search {
        width: 100%;
    }

    .color-green {
        color: limegreen;
    }

    .color-red {
        color: red;
    }

    .icons {
        width: 22px !important;
        height: 22px !important;
    }


</style>