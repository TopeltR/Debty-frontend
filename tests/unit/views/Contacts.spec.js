import Contacts from '@/views/Contacts.vue'
import userStore from '@/stores/UserStore.ts'
import Vue from "vue";
import {mount} from "@vue/test-utils";
import BootstrapVue from "bootstrap-vue";


describe('Contacts', () => {
    Vue.use(BootstrapVue);
    console.error = jest.fn();
    console.warn = jest.fn();

    userStore.getUser = jest.fn(() => Promise.resolve({id: 1}));
    const contactsList = [{id: 2}, {id: 3}];
    const get = jest.fn(() => Promise.resolve({data: contactsList}));
    const post = jest.fn();
    const del = jest.fn();
    Vue.prototype.$http = {get, post, delete: del};

    describe('getContacts', () => {
        it('should get all availableContacts from /contact/all/{id}', async () => {
            const contacts =  mount(Contacts);

            await await contacts.vm.getContacts();

            expect(get).toHaveBeenCalledWith('/contact/all/1');
            expect(contacts.vm.availableContacts).toEqual(contactsList);
        })
    });

    describe('getWaitingContacts', () => {
        it('should get requests from /contact/waiting/{id}', async () => {
            const contacts = mount(Contacts);

            await contacts.vm.getWaitingContacts();

            expect(get).toHaveBeenCalledWith('/contact/waiting/1');
            expect(contacts.vm.requests).toEqual(contactsList.map((c) => ({...c, type: 'Incoming'})));
        })
    });

    describe('getPersonContacts', () => {
        it('should get user contacts from /contact/id/{id}', async () => {
            const contacts = mount(Contacts);

            await contacts.vm.getPersonContacts();

            expect(get).toHaveBeenCalledWith('/contact/id/1');
            expect(contacts.vm.userContacts).toEqual(contactsList)
        })
    });

    describe('addContact', () => {
        it('should post to /contact/add/{user-id}/{contact-id}', async () => {
            const contacts = mount(Contacts);
            contacts.vm.contact = {id: 5};

            await contacts.vm.addContact();

            expect(post).toHaveBeenCalledWith('/contact/add/1/5');
        })
    });

    describe('acceptContact', () => {
        it('should post to /contact/accept/{user-id}/{id}', async () => {
            const contacts = mount(Contacts);

            await contacts.vm.acceptContact(5);

            expect(post).toHaveBeenCalledWith('/contact/accept/1/5')
        })
    });

    describe('removeRequest', () => {
        it('should delete from /contact/remove/{user-id}/{contact-id}', async () => {
            const contacts = mount(Contacts);

            await contacts.vm.removeRequest({id: 5});

            expect(del).toHaveBeenCalledWith('/contact/remove/1/5');
        })
    });

    it('should render', () => {
        const contacts = mount(Contacts);

        expect(contacts.html()).toMatchSnapshot();
    })
});
