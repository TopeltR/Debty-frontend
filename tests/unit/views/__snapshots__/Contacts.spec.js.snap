// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contacts should render 1`] = `
<div>
  <div class="container-fluid">
    <div class="row">
      <nav class="navbar fixed-top w-100 bg-light-gray navbar-light navbar-expand-sm"><button type="button" aria-label="Toggle navigation" aria-controls="nav_collapse" aria-expanded="false" class="navbar-toggler"><span class="navbar-toggler-icon"></span></button>
        <div class="navbar-brand limegreen brand">debty</div>
        <div id="nav_collapse" class="navbar-collapse collapse" style="display: none;">
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/home" class="nav-link">
                Home
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/events" class="nav-link">
                Events
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/debts" class="nav-link">
                Debts
              </a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/contacts" class="nav-link">
                Contacts
                <!----></a></li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item"><a target="_self" href="/profile" class="nav-link">
                Profile
              </a></li>
          </ul>
          <ul class="navbar-nav ml-auto"><span> Welcome, <b></b></span>
            <font-awesome-icon icon="sign-out-alt" class="icon ml-4 mr-2 mt-1"></font-awesome-icon> <span>Sign out</span></ul>
        </div>
      </nav>
    </div>
  </div>
  <div class="diagonal-background container-fluid">
    <div class="container">
      <div class="search-wrapper pt-4">
        <div class="row ml-0">
          <h2>Add contact</h2>
          <font-awesome-icon id="info" icon="info-circle" class="ml-1"></font-awesome-icon>
          <div aria-hidden="true" class="d-none" style="display: none;">
            <div></div>
          </div>
        </div>
        <div class="row pt-4">
          <div class="col-sm-12 col-md-6">
            <div class="autocomplete" id="search " value="[object Object]"><input type="text" placeholder="Name" class="form-control">
              <ul id="autocomplete-results" class="autocomplete-results" style="display: none;"></ul>
            </div>
          </div>
          <div class="pt-3 pt-md-0 col"><button type="button" class="btn btn-primary wide"> Add contact
            </button></div>
        </div>
      </div>
      <div class="row mt-5">
        <div class="col-md-6">
          <div class="list-group">
            <h2>Requests</h2>
            <div class="mt-3">
              <h4>No incoming requests right now.</h4>
            </div>
          </div>
        </div>
        <div class="col-md-6">
          <div class="list-group">
            <h2>My contacts</h2>
            <div class="mt-3">
              <h3>No contacts yet, add contacts from search</h3>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
